apollo-compiler/src/main/kotlin/com/apollographql/apollo3/compiler/backend/codegen/ResponseAdapterReadFromResponse.kt

```diff
@@ -139,6 +155,82 @@ private fun CodeGenerationAst.ObjectType.readFragmentFromResponseFunSpec(): FunS
       .build()
 }
 
+private fun CodeGenerationAst.ObjectType.readFragmentFromBufferedResponseFunSpec(): FunSpec {
```

Comment:
Sounds like this is duplicating a lot of the ResponseAdapter codegen. Would it be possible to somehow factor some stuff?

Reply:
not sure about that, even it looks like code is duplicated it's different `readFragmentFromStreamResponseFunSpec` and `readFragmentFromBufferedResponseFunSpec` emit entirely different code to for reading from response. The only place where they overlap is function declaration.

Reply:
Right. 

For the record, Data classes is like

```
      if (__typename in arrayOf("Droid", "Human")) {
        reader.rewind()
        heroDetails = HeroDetailsAdapter.fromResponse(reader)
      }
      if (__typename in arrayOf("Human")) {
        reader.rewind()
        humanDetails = HumanDetailsAdapter.fromResponse(reader)
      }
```

Interfaces is like

```
      return when(typename) {
        "Droid" -> CharacterHeroAdapter.fromResponse(reader, typename)
        "Human" -> CharacterHumanHeroAdapter.fromResponse(reader, typename)
        else -> OtherHeroAdapter.fromResponse(reader, typename)
      }
```

The main change is how we iterate, in Data classes, we iterate on each type condition while in Interfaces we iterate on each Type case. So agreed. We could factor in some stuff like reading the typename and function declaration but it might not be worth spreading the code in multiple functions